# TD 1 | Introduction à Python pour l'analyse de données

---

Objectifs du TD :

* découvrir Python
* se familiariser avec le langage et le notebook Jupyter
* découvrir et maîtriser les bases des librairies de calcul numérique et d'analyse de données numpy et pandas

---

##  Installation de Python et Jupyter

### Windows

#### Installation Python

Vérifiez si vous avez déjà une version de python installée en ouvrant l'invite de commande et en exécutant l'une des deux commandes python suivantes : 
> python --version   

> python3 --version 

Si ce n'est pas le cas, vous pouvez procéder à l'installation comme suit :
 
1. Allez sur le site web : https://www.python.org/downloads/windows/
2. Cliquez sur "Latest Python 3 Release - Python 3.10.7".
3. Si votre ordinateur utilise une version 64 bits de Windows, téléchargez le Windows installer (64-bit). Sinon, téléchargez Windows installer (32-bit).
4. Après avoir téléchargé le Windows installer, vous devez l'exécuter (double-cliquez dessus) et suivre les instructions qui s'y trouvent.

NB : Au début de l'installation, assurez-vous de cocher la case "Add Python 3.10 to PATH" ou "Add Python to your environment variables" avant de cliquer sur "Install Now".


#### Installation Jupyter

Ouvrir l'invite de commande et exécuter la commande suivante en remplaçant Pseudo_de_votre_user par le nom de l'utilisateur de votre session : 
> cd "C:\Users\Pseudo_de_votre_user\AppData\Local\Programs\Python\Python310\Scripts"

Ensuite, executer la commande suivante :
> pip install jupyterlab

Une fois l'installation terminée avec succès, vous pouvez lancer JupyterLab avec la commande suivante : 
> jupyter-lab


### Linux & Mac OS

#### Installation Python

Normalement, Python est préinstallé sur la quasi-totalité des distributions. Vous n'avez pas besoin de le réinstaller, mais vous pouvez mettre à jour la version si vous le souhaitez.

Vérifiez la version Python installée en exécutant l'une des deux commandes python suivantes sur votre terminal : 
> python --version   

> python3 --version 

#### Installation Jupyter

Vérifier que la commande pip3 est installée. Sinon, l'installer (exemple pour Debian/Ubuntu/Linux Mint, à adapter selon votre distribution) :

> sudo apt install python3-pip

Ensuite, installer JupyterLab en executant la commande suivante:
> pip3 install -U jupyterlab

Enfin, placez-vous dans votre dossier de travail et lancez un notebook :

> jupyter-lab

## Syntaxe et Concepts de base

In [86]:
a = 33
b = 9
c = 2.0

In [87]:
a+b

42

In [88]:
print(a+b)

42


In [89]:
type(a)

int

In [90]:
type(c)

float

In [91]:
a = 'hello'
type(a)

str

In [92]:
# Affiche les entiers pairs de 1 à 10
for i in range(1, 11):
    if i % 2 == 0:
        print(i)

2
4
6
8
10


### List

In [93]:
l = [1,2,3,4,4]

#### len, max, min

In [94]:
print('longueur de la liste =', len(l))
print("Valeur max de l =", max(l))
print("Valeur min de l =", min(l))

longueur de la liste = 5
Valeur max de l = 4
Valeur min de l = 1


#### Ajout d'un élément d'une liste

In [95]:
l.append(6)
l

[1, 2, 3, 4, 4, 6]

In [96]:
l = l+[7]
l

[1, 2, 3, 4, 4, 6, 7]

In [97]:
l+=[7]
l

[1, 2, 3, 4, 4, 6, 7, 7]

#### Modifier un élément

In [98]:
l[0] = 29
l

[29, 2, 3, 4, 4, 6, 7, 7]

In [99]:
l.insert(1,9)
l

[29, 9, 2, 3, 4, 4, 6, 7, 7]

#### Suppression des éléments d'une liste

In [100]:
l.remove(7)
l

[29, 9, 2, 3, 4, 4, 6, 7]

In [101]:
l.pop(2)
l

[29, 9, 3, 4, 4, 6, 7]

#### Trier une liste

In [102]:
l.sort(reverse=True)
print(l)

[29, 9, 7, 6, 4, 4, 3]


In [103]:
l_sorted = sorted(l)
print('Sorted list:', l_sorted)

Sorted list: [3, 4, 4, 6, 7, 9, 29]


#### Liste non homogène

In [104]:
l = [1, 'un', 2]
l.append('deux')
l += [3, 'trois']
print(type(l))
l

<class 'list'>


[1, 'un', 2, 'deux', 3, 'trois']

In [105]:
a = l[3], l[-1]
print(type(a))
a

<class 'tuple'>


('deux', 'trois')

In [106]:
a[0]

'deux'

In [107]:
[2**p for p in range(10)] # "Compréhension de liste"

[1, 2, 4, 8, 16, 32, 64, 128, 256, 512]

In [108]:
# Liste des entiers pairs de 1 à 10
l = [i for i in range(1, 11) if i % 2 == 0]
print(l)

[2, 4, 6, 8, 10]


### Tuple

In [109]:
p = tuple(range(1,5))
print("p=", p)
print("p[0] = ", p[0])

p= (1, 2, 3, 4)
p[0] =  1


In [110]:
p = ([1,2],5)

print(p)

p[0][0] = 8

print(p)

([1, 2], 5)
([8, 2], 5)


In [111]:
# Les éléments du tuple sont fixes
p[0] = 6

TypeError: 'tuple' object does not support item assignment

In [143]:
p1 = (1,2)
p2 = (2,1)

print(p1 == p2)

False


### Set

In [144]:
s = set((1,2,'l'))
print(s)

s = {4,5,9,9,9,9}
print(s)

{'l', 1, 2}
{9, 4, 5}


In [145]:
#Ajout d'un élément au set
s.add(6)
print(s)

{9, 4, 5, 6}


In [146]:
#Suppression d'un élément du set
s.discard(5)
print(s)

{9, 4, 6}


In [147]:
s.clear()
print(s)

set()


In [148]:
A = {1,2,3,4,5}
B = {2,4,6,7}
C = {1,2,3}

In [149]:
#Intersection
print("Intersection de A et B =", A&B)

#Union
print("Union entre A et B =", A|B)

#Exclusion
print("Exclusion de B de A =", A-B)

#Différence symétrique
print("Diff sym. A et B =", A^B)

Intersection de A et B = {2, 4}
Union entre A et B = {1, 2, 3, 4, 5, 6, 7}
Exclusion de B de A = {1, 3, 5}
Diff sym. A et B = {1, 3, 5, 6, 7}


In [150]:
print("B est inclus strict dans A") if A>B else print(False)
print("C est inclus dans A") if A>=C else print(False)

False
C est inclus dans A


### Dictionary

In [151]:
# 1ère façon de créer un dictionnaire
dic = {"key1": "value1", "answer": 42}
# 2ème façon de créer un dictionnaire
dic2 = dict(key1="value1", answer=42)
# 3ème façon de créer un dictionnaire
dic3 = dict([("key1","value1"), ("answer",42)])
print(dic)
print(dic2)
print(dic3)

dic["answer"]

{'key1': 'value1', 'answer': 42}
{'key1': 'value1', 'answer': 42}
{'key1': 'value1', 'answer': 42}


42

In [152]:
dic['new'] = [1,2,3]

In [153]:
dic.keys()

dict_keys(['key1', 'answer', 'new'])

In [154]:
dic.items()

dict_items([('key1', 'value1'), ('answer', 42), ('new', [1, 2, 3])])

In [155]:
dic.values()

dict_values(['value1', 42, [1, 2, 3]])

In [156]:
a,b,c = dic.items()
print(a)

('key1', 'value1')


## Programmation Orientée Objet

In [157]:
"""
Exemple de classe en Python
"""
class Moteur:
    
    # Constructeur
    def __init__(self, esn, panne=False):
        self.esn = esn
        self.panne = panne
    
    # Méthodes
    def dire_bonjour(self):
        print('Bonjour, mon numéro de série est ' + self.esn)
    
    def fonctionne(self):
        return not self.panne 
    

In [158]:
mot1 = Moteur('420912')
mot2 = Moteur(panne=True, esn='420913')

mot1.dire_bonjour()
print(mot1.fonctionne())

mot1.panne = True
print(mot1.fonctionne())

print('\n')
mot2.dire_bonjour()
print(mot2.fonctionne())

Bonjour, mon numéro de série est 420912
True
False


Bonjour, mon numéro de série est 420913
False


### Méthode récurcive vs Méthode itérative

In [159]:
# Méthode récursive
def fact1(n):
    if n == 0:
        return 1
    else:
        return n*fact1(n-1)

# Méthode itérative
def fact2(n):
    s=1
    for i in range(1,n+1):
        s*=i
    return s

print(fact1(5))
print(fact2(5))

120
120


In [160]:
"""
App.1 : Ecrire une fonction qui retourne le plus grand commun diviseur (pgcd) entre deux nombres a et b.
"""
def pgcd(a,b):
    s=b
    while s!=0:
        M=max(a,b)
        m=min(a,b)
        s=M-m 
        a=min(a,b)
        b=s
    return a
print(pgcd(25,230))
            

5


In [161]:
"""
App.1 : Correction
"""

#Méthode itérative
def pgcd1(a,b):
    while (a%b) != 0:
        s = a%b
        a = b
        b = s
    return b


#Méthode récurcive
def pgcd2(a,b):
    if a==b : return a 
    if a<b : a,b = b,a
    return pgcd2(a-b,b)

a = 230
b = 25

print("pgcd1(a,b) = ", pgcd1(a,b))
print("pgcd2(a,b) = ", pgcd2(a,b))

pgcd1(a,b) =  5
pgcd2(a,b) =  5


## Libraries

> pip install numpy pandas

> pip3 install -U numpy pandas

### Math

In [162]:
import math

In [163]:
x = 4

print("sqrt(x) =", math.sqrt(x))
print("cos(x) =", math.cos(x))
print("sin(x) =", math.sin(x))
print("factorial(x) =", math.factorial(x))

sqrt(x) = 2.0
cos(x) = -0.6536436208636119
sin(x) = -0.7568024953079282
factorial(x) = 24


### Numpy

In [164]:
import numpy as np

In [165]:
# Création d'un array à partir d'une liste
v = np.array([1.0, 2.0, 3.0])
# Création d'un array de taille (n,m) initialisé à 0
z = np.zeros((3,4))
z

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [166]:
# Taille d'un array
print(z.shape)
print(v.shape)

(3, 4)
(3,)


In [167]:
# Opérations courantes
A = np.ones((3,3)) + np.eye(3)
print('A = ', A)
print('Av = ', np.dot(A,v))
print('3*A = ', 3*A)
print('A*v = ', A*v)
print('A + 1 = ', A+1)
print('A + v = ', A+v)
print('A^2 = ', np.square(A))

A =  [[2. 1. 1.]
 [1. 2. 1.]
 [1. 1. 2.]]
Av =  [7. 8. 9.]
3*A =  [[6. 3. 3.]
 [3. 6. 3.]
 [3. 3. 6.]]
A*v =  [[2. 2. 3.]
 [1. 4. 3.]
 [1. 2. 6.]]
A + 1 =  [[3. 2. 2.]
 [2. 3. 2.]
 [2. 2. 3.]]
A + v =  [[3. 3. 4.]
 [2. 4. 4.]
 [2. 3. 5.]]
A^2 =  [[4. 1. 1.]
 [1. 4. 1.]
 [1. 1. 4.]]


In [168]:
B = np.array([5, 2, 9, 1])

print(np.sort(B))

[1 2 5 9]


#### Compatif de performance : produit matriciel


In [169]:
import random

In [333]:
# Création d'une matrice aléatoire sous forme de liste de listes
taille = 2
A = [[random.random() for _ in range(taille)] for _ in range(taille)]

In [216]:
print(A)

[[0.14363961672842251, 0.6415546711367116], [0.9865247997857114, 0.7058417540335712]]


In [172]:
"""
EXERCICE - Afficher un tuple contenant les dimensions de la matrice A
"""
dim=np.shape(A)
print(dim)
len(A)

(2, 2)


2

$$ (AB)_{ij} = \sum_k A_{ik} B_{kj} $$

In [217]:
"""
EXERCICE - Implémenter le produit matriciel de 2 matrices sous forme de listes de listes python
"""
def produit(A,B):
    dimA=np.shape(A)
    dimB=np.shape(B)
    if dimA[1]!= dimB[0]:
        print("erreur de dimensions des matrices")
        return 0
    P=np.zeros((dimA[0], dimB[1]))
    for i in range(dimA[0]):
        for j in range(dimB[1]):
            S=0
            for k in range(dimA[1]):
                S=S+A[i][k]*B[k][j]
            P[i][j]=S 
    return P




    

NameError: name 'P' is not defined

In [228]:
# Vérification
res=(produit([[1, 2], [3, 4]], [[1, 2], [3, 4]]) == [[7, 10], [15, 22]]).all()
print(res)
# assert(produit([[1, 2], [3, 4]], [[1, 2], [3, 4]]) == [[7, 10], [15, 22]])

True


In [334]:
P=produit(A, A)
type(P)

numpy.ndarray

In [229]:
# Sortons le chronomètre
%timeit produit(A, A)

10.3 µs ± 426 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [230]:
# Et maintenant, avec numpy !
import numpy as np

In [231]:
A2 = np.array(A)
print(A2)

[[0.14363962 0.64155467]
 [0.9865248  0.70584175]]


In [232]:
%timeit np.dot(A, A)

5.3 µs ± 273 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [237]:
A=np.random.rand(30,30)
%timeit np.dot(A,A)
%timeit produit(A,A)
A=np.random.rand(300,300)
%timeit np.dot(A,A)
%timeit produit(A,A)
A=np.random.rand(3000,3000)
%timeit np.dot(A,A)

5.16 µs ± 91.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
17.9 ms ± 751 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
894 µs ± 61.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
17.4 s ± 326 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
553 ms ± 34.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


EXERCICE - Remplissez le tableau suivant avec les durées d'exécution constatées :

(Conseil : pour la taille 3000, essayez UNIQUEMENT avec numpy)

Taille | Python | numpy
-------|--------|-------
30     |  17.9 ms ± 751 µs  | 5.16 µs ± 91.7
300    |  17.4 s ± 326 ms   | 894 µs ± 61.1 µs
3000   |        XXX         | 553 ms ± 34.2 

###  Pandas

In [238]:
import pandas as pd

In [239]:
df_exemple = pd.DataFrame({"ESN": ["E420912", "E420913", "E420914"], "panne": [False, False, True]})
df_exemple

ESN  panne
0  E420912  False
1  E420913  False
2  E420914   True

#### Lecture et prétraitement de données


In [240]:
!git clone https://github.com/MadaneA/MACS3-Statistiques-Descriptives-TDs.git

Cloning into 'MACS3-Statistiques-Descriptives-TDs'...


In [241]:
import os
os.chdir('MACS3-Statistiques-Descriptives-TDs')

In [242]:
 os.getcwd()

'c:\\Users\\djame\\Desktop\\MACS 3\\Stats descriptives\\TD_sd\\MACS3-Statistiques-Descriptives-TDs'

In [306]:
# Chargement d'un fichier CSV ou Excel
df = pd.read_csv("./data/Vol010.csv")
# Affichage des 5 premières lignes
df.head()

t EGT_SEL FLIGHT_MOD FMV_SEL HPTC_SEL LPTC_SEL  \
0                  datenum   deg_C          _       _        %        %   
1  15/09/2011 14:25:58.125     NaN        NaN     NaN      NaN      NaN   
2  15/09/2011 14:25:58.375     NaN        NaN     NaN      NaN      NaN   
3  15/09/2011 14:25:58.625     NaN        NaN     NaN      NaN      NaN   
4  15/09/2011 14:25:58.875     NaN        NaN     NaN      NaN      NaN   

  N1_SEL N2_ACTSEL OIL_P OIL_TEMP  ... T25_SEL T3_SEL VBV_SEL VIB_CN1 VIB_CN2  \
0      %     %_RPM   psi        _  ...       _      _     DEG       _       _   
1    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   
2    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   
3    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   
4    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   

  VIB_TN1 VIB_TN2 VSV_SEL  WFM_SEL    XM  
0       _       _     DEG     lb/h  mach  
1     NaN     NaN     NaN      NaN   NaN  
2     NaN     NaN     NaN      NaN   NaN  
3     NaN     NaN     NaN      NaN   NaN  
4     NaN     NaN     NaN  7679.98  0.15  

[5 rows x 26 columns]

In [307]:
"""
EXERCICE - Dimensions d'un DataFrame
Affichez le nombre de colonnes et de ligne du DataFrame (indice : beaucoup de méthodes sont communes entre numpy et pandas)
"""
dim=df.shape
print('Nombre de colonnes :', dim[1])
print('Nombre de lignes :', dim[0])

Nombre de colonnes : 26
Nombre de lignes : 22945


In [308]:
df.size, len(df)

(596570, 22945)

In [309]:
df.columns

Index(['t', 'EGT_SEL', 'FLIGHT_MOD', 'FMV_SEL', 'HPTC_SEL', 'LPTC_SEL',
       'N1_SEL', 'N2_ACTSEL', 'OIL_P', 'OIL_TEMP', 'PS3_SEL', 'PT2_SEL',
       'P0_SEL', 'TAT', 'TBV_SEL', 'TRA_SEL', 'T25_SEL', 'T3_SEL', 'VBV_SEL',
       'VIB_CN1', 'VIB_CN2', 'VIB_TN1', 'VIB_TN2', 'VSV_SEL', 'WFM_SEL', 'XM'],
      dtype='object')

Les colonnes des DF sont typées, à la manière d'une base de données relationnelle, contrairement aux variables python classiques. Les types des colonnes sont accessibles via df.dtypes. Les principaux types sont les numériques (int32, int64, float etc.

In [310]:
df.loc[10:15, ['EGT_SEL', 'FLIGHT_MOD']]

EGT_SEL FLIGHT_MOD
10    0.00234327        NaN
11  1.08032e-007        NaN
12  3.12514e-012        NaN
13             0        NaN
14       72.0006        NaN
15       215.996        NaN

In [311]:
"""
EXERCICE - Extraction et suppression des unités
On remarque que la 1ère ligne ne contient pas de données mais les unités de chaque colonne.
Pour la suite des traitements, il faut supprimer cette ligne. On souhaite toutefois garder l'information des unités de chaque colonne.
1. Récupérez les unités et stockez les dans une structure adaptée.
2. Supprimez cette ligne du DataFrame en utilisant la méthode "drop"
"""
units=list(df.loc[0,])
df=df.drop(0)

In [312]:
print(units)
df.head()

['datenum', 'deg_C', '_', '_', '%', '%', '%', '%_RPM', 'psi', '_', 'psi', 'mb', 'psi', 'deg_C', '%', 'DEG', '_', '_', 'DEG', '_', '_', '_', '_', 'DEG', 'lb/h', 'mach']


t EGT_SEL FLIGHT_MOD FMV_SEL HPTC_SEL LPTC_SEL  \
1  15/09/2011 14:25:58.125     NaN        NaN     NaN      NaN      NaN   
2  15/09/2011 14:25:58.375     NaN        NaN     NaN      NaN      NaN   
3  15/09/2011 14:25:58.625     NaN        NaN     NaN      NaN      NaN   
4  15/09/2011 14:25:58.875     NaN        NaN     NaN      NaN      NaN   
5  15/09/2011 14:25:59.125     NaN        NaN     NaN      NaN      NaN   

  N1_SEL N2_ACTSEL    OIL_P OIL_TEMP  ... T25_SEL T3_SEL VBV_SEL VIB_CN1  \
1    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
2    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
3    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
4    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
5      0         0  647.998      NaN  ...     NaN    NaN       0       0   

  VIB_CN2 VIB_TN1 VIB_TN2 VSV_SEL  WFM_SEL    XM  
1     NaN     NaN     NaN     NaN      NaN   NaN  
2     NaN     NaN     NaN     NaN      NaN   NaN  
3     NaN     NaN     NaN     NaN      NaN   NaN  
4     NaN     NaN     NaN     NaN  7679.98  0.15  
5       0       0       0       0  4607.86  0.15  

[5 rows x 26 columns]

On remarque que toutes les colonnes ont été reconnues comme de type object, c'est-à-dire des chaînes de caractères, alors que ce sont des valeurs numériques. Cela est dû à la première ligne contenant les unités. Il faut donc convertir les colonnes en numérique. La colonne 't', quant à elle, doit être convertie en type datetime.

In [313]:
df['t'] = pd.to_datetime(df['t'])
df[df.columns[1:]] = df[df.columns[1:]].apply(pd.to_numeric)

C:\Users\djame\AppData\Local\Temp\ipykernel_9588\2306591669.py:1: UserWarning: Parsing dates in %d/%m/%Y %H:%M:%S.%f format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['t'] = pd.to_datetime(df['t'])


In [314]:
df.dtypes

t             datetime64[ns]
EGT_SEL              float64
FLIGHT_MOD           float64
FMV_SEL              float64
HPTC_SEL             float64
LPTC_SEL             float64
N1_SEL               float64
N2_ACTSEL            float64
OIL_P                float64
OIL_TEMP             float64
PS3_SEL              float64
PT2_SEL              float64
P0_SEL               float64
TAT                  float64
TBV_SEL              float64
TRA_SEL              float64
T25_SEL              float64
T3_SEL               float64
VBV_SEL              float64
VIB_CN1              float64
VIB_CN2              float64
VIB_TN1              float64
VIB_TN2              float64
VSV_SEL              float64
WFM_SEL              float64
XM                   float64
dtype: object

In [315]:
"""
Exercice - Index temporel
Comme nos données sont une série temporelle multivariée, on souhaite utiliser un index temporel.
1. Créez une copie de df, appelée df2, à l'aide de la méthode du même nom.
2. Affectez la colonne du temps ('t') en tant qu'indice du DataFrame.
3. Supprimez la colonne 't' du DF résultant.
"""
df2=df
df2.set_index('t', inplace= True)
df2.head()



EGT_SEL  FLIGHT_MOD  FMV_SEL  HPTC_SEL  LPTC_SEL  \
t                                                                           
2011-09-15 14:25:58.125      NaN         NaN      NaN       NaN       NaN   
2011-09-15 14:25:58.375      NaN         NaN      NaN       NaN       NaN   
2011-09-15 14:25:58.625      NaN         NaN      NaN       NaN       NaN   
2011-09-15 14:25:58.875      NaN         NaN      NaN       NaN       NaN   
2011-09-15 14:25:59.125      NaN         NaN      NaN       NaN       NaN   

                         N1_SEL  N2_ACTSEL    OIL_P  OIL_TEMP  PS3_SEL  ...  \
t                                                                       ...   
2011-09-15 14:25:58.125     NaN        0.0      NaN       NaN      NaN  ...   
2011-09-15 14:25:58.375     NaN        0.0      NaN       NaN      NaN  ...   
2011-09-15 14:25:58.625     NaN        0.0      NaN       NaN      NaN  ...   
2011-09-15 14:25:58.875     NaN        0.0      NaN       NaN      NaN  ...   
2011-09-15 14:25:59.125     0.0        0.0  647.998       NaN  575.982  ...   

                         T25_SEL  T3_SEL  VBV_SEL  VIB_CN1  VIB_CN2  VIB_TN1  \
t                                                                              
2011-09-15 14:25:58.125      NaN     NaN      NaN      NaN      NaN      NaN   
2011-09-15 14:25:58.375      NaN     NaN      NaN      NaN      NaN      NaN   
2011-09-15 14:25:58.625      NaN     NaN      NaN      NaN      NaN      NaN   
2011-09-15 14:25:58.875      NaN     NaN      NaN      NaN      NaN      NaN   
2011-09-15 14:25:59.125      NaN     NaN      0.0      0.0      0.0      0.0   

                         VIB_TN2  VSV_SEL  WFM_SEL    XM  
t                                                         
2011-09-15 14:25:58.125      NaN      NaN      NaN   NaN  
2011-09-15 14:25:58.375      NaN      NaN      NaN   NaN  
2011-09-15 14:25:58.625      NaN      NaN      NaN   NaN  
2011-09-15 14:25:58.875      NaN      NaN  7679.98  0.15  
2011-09-15 14:25:59.125      0.0      0.0  4607.86  0.15  

[5 rows x 25 columns]

On constate que pandas a automatiquement reconnu un DatetimeIndex, adapté pour des manipulations de séries temporelles (moyennes glissantes, etc) !

### Valeurs manquantes

**NaN = Not a Number**

Les valeurs NaN doivent être éliminées ou imputées (i.e. remplacées par une certaine valeur) avant la suite des traitements. Ce choix dépend du cas d'usage. Dans un premier temps, nous allons apprendre à :

* trouver les données manquantes (méthode `isna`)
* éliminer les données manquantes d'un DataFrame (méthode `dropna`)
* les remplacer par une constante (méthode (`fillna`)

In [429]:
"""
EXERCICE - La méthode isna
1. Testez la méthode isna sur le DataFrame df2, puis sur une colonne ou une ligne. Que renvoie-t-elle ?
2. En appliquant les méthodes any(axis=...), mean() et max()/idxmax() sur les résultats de isna(), répondez aux questions suivantes :
    2.1 Quelles colonnes contiennent des valeurs manquantes, lesquelles n'en contiennent pas ?
    2.2 Quel est le pourcentage de valeurs manquantes dans le DF (a) par colonne (b) globalement ? Quelle variable contient le plus de NaN ?
    2.3 Quel est le pourcentage d'indices du DF pour lesquels toutes les variables sont présentes ?
"""
valeurs_manquantes=df2.isna()
"""
1. La méthode isna renvoie une matrice (ou vecteur) de booléen True/False indiquant si la valeur située à la position est manquante (True) ou non (False)
"""
colonnes_true=valeurs_manquantes.any(axis=0)
print('Question 2.1')
print('les colonnes qui ont des valeurs manquantes sont : \n',list(dict(colonnes_true).keys()))
print(f"il y a exactement",len(df2.columns)-len(colonnes_true), "colonnes sans aucunes valeurs manquantes")
print('Question 2.2')
pct_colonnes=valeurs_manquantes.mean(axis=0)
pct_lignes=valeurs_manquantes.mean(axis=1)
print("les pourcentages de valeurs manquantes par colonnes sont: \n", dict(pct_colonnes*100))
print("le pourcentage de valeurs manquantes globablement est:", valeurs_manquantes.values.sum()/valeurs_manquantes.size*100)
print("la variable qui contient le plus de NaN est:", pct_colonnes.idxmax())
print('Question 2.3')
print("le pourcentage d'indice du DF ayant toutes les variables présente est:", (pct_lignes==0).mean()*100)

Question 2.1
les colonnes qui ont des valeurs manquantes sont : 
 ['EGT_SEL', 'FLIGHT_MOD', 'FMV_SEL', 'HPTC_SEL', 'LPTC_SEL', 'N1_SEL', 'N2_ACTSEL', 'OIL_P', 'OIL_TEMP', 'PS3_SEL', 'PT2_SEL', 'P0_SEL', 'TAT', 'TBV_SEL', 'TRA_SEL', 'T25_SEL', 'T3_SEL', 'VBV_SEL', 'VIB_CN1', 'VIB_CN2', 'VIB_TN1', 'VIB_TN2', 'VSV_SEL', 'WFM_SEL', 'XM']
il y a exactement 0 colonnes sans aucunes valeurs manquantes
Question 2.2
les pourcentages de valeurs manquantes par colonnes sont: 
 {'EGT_SEL': 0.3486750348675035, 'FLIGHT_MOD': 0.444560669456067, 'FMV_SEL': 0.444560669456067, 'HPTC_SEL': 0.42276847977684795, 'LPTC_SEL': 0.45763598326359833, 'N1_SEL': 0.41841004184100417, 'N2_ACTSEL': 0.25278940027894004, 'OIL_P': 0.37918410041841005, 'OIL_TEMP': 0.38354253835425384, 'PS3_SEL': 0.30073221757322177, 'PT2_SEL': 0.3704672245467225, 'P0_SEL': 0.3704672245467225, 'TAT': 0.30073221757322177, 'TBV_SEL': 0.29637377963737793, 'TRA_SEL': 0.30073221757322177, 'T25_SEL': 0.47071129707112974, 'T3_SEL': 0.470711297071

In [393]:
"""
EXERCICE - La méthode dropna
La méthode dropna permet d'éliminer les valeurs manquantes (NaN). Lisez d'abord sa documentation.
1. À quoi correspondent les arguments "axis" et "how" ?
2. Éliminez toutes les lignes contenant uniquement des valeurs manquantes.
3. Éliminez toutes les lignes contenant au moins une valeur manquante. Combien y a-t-il de lignes de différence ?
4. Éliminez toutes les colonnes contenant au moins une valeur manquante.
"""
"""
Question 1
L'argument axis correspond à l'axe de la suppression. 0 pour supprimer en lignes et 1 pour supprimer en colonnes. L'argument how correspond au critère de la suppression. Any pour supprimer toutes les lignes/colonnes ayant au monis une valeur manquante et all pour supprimer celles qui n'ont que des valeurs manquantes.

"""
print("Question 2")
df2_dropall=df2.dropna(axis=0, how='all')
print("Question 3")
df2_dropany=df2.dropna(axis=0, how='any')
print(f"Il y a ", len(df2_dropall)-len(df2_dropany), "lignes de plus dans le df2_dropall")
print("Question 4")
df2_dropcol=df2.dropna(axis=1, how='any')

Question 2
Question 3
Il y a  64 lignes de plus dans le df2_dropall
Question 4


In [422]:
"""
EXERCICE - La méthode fillna
La méthode dropna permet d'imputer les valeurs manquantes (NaN). Lisez d'abord sa documentation.
1. Quelles sont les différentes stratégies de remplissage des valeurs manquantes ?
2. Imputez les valeurs manquantes de la colonne age du DF donné en exemple par :
    - 0
    - la dernière valeur précédente/suivante valide
    - la moyenne
    - la valeur la plus courante (mode)
3. Quel est le meilleur choix dans ce cas ? Et pour le cas d'une variable temporelle, par exemple la température 'EGT_SEL' de notre jeu de données ?
"""
exemple = pd.DataFrame({'nom': ['Alice', 'Bob', 'Charlie', 'David'], 'age': [24, np.nan, 99, 24]})
print(exemple,"\n")
"""
Question 1
On peut imputer les valeurs manquantes par n'importe quelle valeur fixée, par la valeur précédente ou suivante, par la moyenne, par la médiane ou par la valeur la plus courante.

"""
fill_0=exemple.fillna(value=0)
fill_pad=exemple.fillna(method='ffill')
fill_mean=exemple.fillna(exemple['age'].mean())
fill_mode=exemple.fillna(exemple['age'].mode()[0])

print("Question 2 \n -0:\n", fill_0,"\n-Valeur précédente:\n", fill_pad,"\n -Moyenne:\n", fill_mean, "\n-mode:\n", fill_mode)

"""
Question 3 
Le meilleur choix est la moyenne ou la valeur la plus présente dans ce cas. Dans le cas d'une variable temporelle, afin de garder un cohérence le meilleur choix est la valeur précédente ou la valeur suivante. 

"""


       nom   age
0    Alice  24.0
1      Bob   NaN
2  Charlie  99.0
3    David  24.0 

Question 2 
 -0:
        nom   age
0    Alice  24.0
1      Bob   0.0
2  Charlie  99.0
3    David  24.0 
-Valeur précédente:
        nom   age
0    Alice  24.0
1      Bob  24.0
2  Charlie  99.0
3    David  24.0 
 -Moyenne:
        nom   age
0    Alice  24.0
1      Bob  49.0
2  Charlie  99.0
3    David  24.0 
-mode:
        nom   age
0    Alice  24.0
1      Bob  24.0
2  Charlie  99.0
3    David  24.0


C:\Users\djame\AppData\Local\Temp\ipykernel_9588\605604680.py:20: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  fill_pad=exemple.fillna(method='ffill')


"\nQuestion 3 \nLe meilleur choix est la moyenne ou la valeur la plus présente dans ce cas. Dans le cas d'une variable temporelle, afin de garder un cohérence le meilleur choix est la valeur précédente ou la valeur suivante. \n\n"

## Git

Créer un compte GitHub : https://github.com/

Créer un nouveau repository :
1. Sélectionner "Nouveau dépôt" dans le menu déroulant avec le signe +. 
2. Saisisser un nom pour votre dépôt (par exemple, "TDs statistiques descriptives") 
3. Cliquer sur "Créer un dépôt". Ne vous souciez pas des autres options.



Télécharger Git Bash : https://git-scm.com/downloads

Sur Git Bash, positionnez-vous dans le répertoire où votre projet figure.

Ensuite

> git init

> git remote add origin https://github.com/********

> git add .

> git commit -m "Ajout du TD1"

> git push origin master

C'est tout ! Vous avez créé votre repo GitHub ! 